app/uk/gov/hmrc/agentclientrelationships/repository/AgentReferenceRepository.scala

```
@@ -0,0 +1,103 @@
+  def findByArn(arn: Arn): Future[Option[AgentReferenceRecord]]
+  def updateAgentName(uid: String, newAgentName: String): Future[Unit]
+  def delete(arn: Arn): Future[Unit]
+}
+
+@Singleton
+class MongoAgentReferenceRepository @Inject() (mongo: MongoComponent)(implicit ec: ExecutionContext)
+    extends PlayMongoRepository[AgentReferenceRecord](
+      mongoComponent = mongo,
+      collectionName = "agent-reference",
+      domainFormat = formats,
+      indexes = List(
+        IndexModel(ascending("uid"), IndexOptions().unique(true)),
+        IndexModel(ascending("arn"), IndexOptions().unique(true))
+      )
+    )
+    with AgentReferenceRepository
+    with Logging {
+
+  override lazy val requiresTtlIndex: Boolean = false
+
+  override def create(agentReferenceRecord: AgentReferenceRecord): Future[Unit] =
+    collection
+      .insertOne(agentReferenceRecord)
+      .toFuture()
+      .map(_ => ())
+
+  override def findBy(uid: String): Future[Option[AgentReferenceRecord]] =
+    collection
+      .find(equal("uid", uid))
+      .headOption()
+
+  override def findByArn(arn: Arn): Future[Option[AgentReferenceRecord]] =
+    collection
+      .find(equal("arn", arn.value))
+      .headOption()
+
+  override def updateAgentName(uid: String, newAgentName: String): Future[Unit] =
+    collection
+      .updateOne(equal("uid", uid), addToSet("normalisedAgentNames", newAgentName), UpdateOptions().upsert(false))
+      .toFuture()
+      .map { updateOneResult =>
+        if (updateOneResult.getMatchedCount == 1L && updateOneResult.getModifiedCount == 1L) ()
+        else throw new RuntimeException("could not update agent reference name, no matching uid found.")
+      }
+
+  override def delete(arn: Arn): Future[Unit] =
+    collection
+      .deleteOne(equal("arn", arn.value))
+      .toFuture()
+      .map { r =>
+        if (r.getDeletedCount == 1L) ()
+        else throw new RuntimeException("could not delete agent reference record, no matching ARN found.")
```

Do we need to throw an exception here? can't we be idempotent and return success if it couldn't find the record to delete?  We can still log though for our records.

---

Agreed - I will change